File: dragoncpp/dragon.cpp
#include <iostream>
#include <cstdio>
#include <queue>

class Dragao
{
public:
    int dias, multa;
    double prioridade;

    Dragao(int d, int m)
    {
        dias = d;
        multa = m;
        prioridade = (double) multa / dias;
    }

    bool operator <(const Dragao& a) const
    {
        return prioridade < a.prioridade;
    }
};

int main(void)
{
    int tempo, multa;
    double multadia = 0.0; // Multa paga por dia por todos os dragões
    double multatotal = 0.0; // Resposta
    bool treinando = false;
    int diasrestantes; // Tempo restante do treinamento
    std::priority_queue<Dragao> fila;

    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    while (std::cin >> tempo >> multa)
    {
        fila.push(Dragao(tempo, multa));
        multadia += multa;
        
        if (!treinando || !diasrestantes)
        {
            multadia -= fila.top().multa;
            diasrestantes = fila.top().dias;
            fila.pop();
            treinando = true;
        }
        multatotal += multadia;
        diasrestantes--;
    }

    if (treinando)
        multatotal +=  diasrestantes * multadia;

    while (!fila.empty()) {
        multadia -= fila.top().multa;
        multatotal += multadia * fila.top().dias;
        fila.pop();
    }

    printf("%.lf\n", multatotal);

    return 0;
}

File: dragonpy/dragon.py
import heapq

class Dragao:
    def __init__(self, dias, multa):
        self.dias = dias
        self.multa = multa
        self.prioridade = multa / dias

    def __lt__(self, other):
        return self.prioridade < other.prioridade

multadia = 0.0  # Multa paga por dia por todos os dragões
multatotal = 0.0  # Resposta
treinando = False
diasrestantes = 0  # Tempo restante do treinamento
fila = []

while True:
    try:
        entrada = input().split()
        tempo, multa = map(int, entrada)
        heapq.heappush(fila, Dragao(tempo, multa))
        multatotal += multa  # Adiciona a multa ao total
        
        if not treinando or not diasrestantes:
            if fila:  # Verifica se há dragões na fila
                multadia += fila[0].multa  # Adiciona a multa do próximo dragão ao total diário
                diasrestantes = fila[0].dias
                heapq.heappop(fila)
                treinando = True

        diasrestantes -= 1
        multatotal += multadia * (diasrestantes + 1)  # Adiciona a multa diária ao total
    except EOFError:
        break
    except ValueError:
        print("Entrada inválida. Por favor, insira dois números separados por espaço.")
        continue

while fila:
    multatotal += fila[0].multa * fila[0].dias  # Adiciona a multa acumulada pelos dias de treinamento do dragão restante
    heapq.heappop(fila)

print(int(multatotal))
